Add support for grouping by artist to beatmap carousel v2

peppy opened the pull request:
One of the final pieces to the puzzle – an example of how groups and sets can interact together.

Note that selection behaviour still needs some thought when for instance opening a group via click then hitting left/right arrows, but this isn't any new breakage so I will address separately.

https://github.com/user-attachments/assets/8addf7b7-b6b9-4eb3-82b9-ceac3c18ea3f

peppy committed:
Add support for grouping by artist

bdach commented on osu.Game.Tests/Visual/SongSelect/BeatmapCarouselV2TestScene.cs:
```diff
@@ -193,7 +193,7 @@ protected void ClickVisiblePanel<T>(int index)
         {
             for (int i = 0; i < count; i++)
-                BeatmapSets.Add(TestResources.CreateTestBeatmapSetInfo(fixedDifficultiesPerSet ?? RNG.Next(1, 4)));
+                BeatmapSets.Add(TestResources.CreateTestBeatmapSetInfo(fixedDifficultiesPerSet ?? RNG.Next(1, 4), randomiseMetadata: true));
```
Application of the metadata randomisation here is breaking tests.

bdach commented on osu.Game.Tests/Resources/TestResources.cs:
```diff
@@ -85,7 +85,8 @@ public static string GetTestBeatmapForImport(bool virtualTrack = false)
         /// <param name="rulesets">Rulesets to cycle through when creating difficulties. If <c>null</c>, osu! ruleset will be used.</param>
-        public static BeatmapSetInfo CreateTestBeatmapSetInfo(int? difficultyCount = null, RulesetInfo[] rulesets = null)
+        /// <param name="randomiseMetadata">Whether to randomise metadata to create a better distribution.</param>
+        public static BeatmapSetInfo CreateTestBeatmapSetInfo(int? difficultyCount = null, RulesetInfo[] rulesets = null, bool randomiseMetadata = false)
```
Not too hot on the metadata randomisation reaching all the way out to here. Would prefer doing it locally to the only usage for now.

bdach commented on osu.Game/Screens/SelectV2/BeatmapCarousel.cs:
```diff
@@ -170,13 +162,24 @@ private void setExpansionStateOfGroup(GroupDefinition group, bool expanded)
                         i.IsVisible = expanded;
                 }
+
+                // Second pass to hide set children when not meant to be displayed.
```
Is there any reason this has to be a separate pass?

peppy replied:
It was just to simplify the logical application ordering (in the above loop it handles beatmaps that aren't grouped in a set, then we undo the incorrect ones).

I do foresee needing to change this for performance reasons. If you have an issue with it I can try that in this PR.

bdach replied:
> I do foresee needing to change this for performance reasons. If you have an issue with it I can try that in this PR.

I'm fine with letting you choose whether you want to address this now or later. The question was more to figure out the reasoning than anything.

peppy replied:
I'll fix this as part of this PR, so please hold off review until it's ready 😃

bdach commented:
Additionally, when clicking on a group header when beatmap set headers are also active, the selection doesn't actually change to the first beatmap of the first beatmap set of the group, which leads to you getting scrolled away from the group header that you just clicked:

https://github.com/user-attachments/assets/bf6596ff-93f3-44ec-965f-edcda4f64b28

peppy commented:
> Additionally, when clicking on a group header when beatmap set headers are also active, the selection doesn't actually change to the first beatmap of the first beatmap set of the group, which leads to you getting scrolled away from the group header that you just clicked

Yeah I was leaving this to fix the scroll part later, rather than make it select. In stable it doesn't automatically select so I'm leaning towards keeping that behaviour.

peppy committed:
Move metadata randomisation local to usage

peppy committed:
Expose set grouping state

peppy committed:
Reorganise tests to make more logical when manually testing

peppy committed:
Fix keyboard traversal on a collapsed group not working as intended

peppy committed:
Add test coverage of artist grouping

peppy committed:
Move selected retrieval functions to base class

peppy committed:
Refactor selection and activation handling

I had a bit of a struggle getting header traversal logic to work well.
The constraints I had in place were a bit weird:

- Group panels should toggle or potentially fall into the prev/next
  group
- Set panels should just traverse around them

The current method of using `CheckValidForGroupSelection` return type
for traversal did not mesh with the above two cases. Just trust me on
this one since it's quite hard to explain in words.

After some re-thinking, I've gone with a simpler approach with one
important change to UX: Now when group traversing with a beatmap set
header currently keyboard focused, the first operation will be to reset
keyboard selection to the selected beatmap, rather than traverse.

I find this non-offensive – at most it means a user will need to press
their group traversal key one extra time.

I've also changed group headers to always toggle expansion when doing
group traversal with them selected.

To make all this work, the meaning of `Activation` has changed somewhat.
It is now the primary path for carousel implementations to change
selection of an item. It is what the `Drawable` panels call when they
are clicked.

Selection changes are not performed implicitly by `Carousel` – an
implementation should decide when it actually wants to change the
selection, usually in `HandleItemActivated`.

Having less things mutating `CurrentSelection` is better in my eyes, as
we see this variable as only being mutated internally when utmost
required (ie the user has requested the change). With this change,
`CurrentSelection` can no longer become of a non-`T` type (in the
beatmap carousel implementation at least).

This might pave a path forward for making `CurrentSelection` typed, but
that comes with a few other concerns so I'll look at that as a
follow-up.

peppy committed:
Avoid double iteration when updating group states

peppy commented:
Apologies for the size increase of this PR, but it's going to be hard to split out the new logic (and a lot is tests anyway).

Please read over 024fbde0fd723a721eba48279085e0539bec0dde carefully.

peppy commented:
@frenzibyte could you also check the changes to see if they work okay in your eyes?

frenzibyte commented on osu.Game/Screens/SelectV2/BeatmapCarousel.cs:
```diff
@@ -170,12 +173,46 @@ private void setExpansionStateOfGroup(GroupDefinition group, bool expanded)
+                                i.IsExpanded = true;
+                                break;
+
+                            case BeatmapSetInfo set:
+                                // Case where there are set headers, header should be visible
+                                // and items should use the set's expanded state.
+                                i.IsVisible = true;
+                                setExpansionStateOfSetItems(set, i.IsExpanded);
+                                break;
+
+                            default:
+                                // Case where there are no set headers, all items should be visible.
+                                if (!grouping.BeatmapSetsGroupedTogether)
+                                    i.IsVisible = true;
+                                break;
```
This comment does not mean to block this PR or request changes in any way shape or form, but is basically a general question and fruit for thought.

Seeing this kind of logic really makes me wonder just how bad would it really be if `CarouselItem` acts as a basic tree model and this logic becomes recursive instead, removing the need for `BeatmapSetsGroupedTogether` or manual lookup of the containing group in which a beatmap is at.

I can imagine multiple things being tackled with this idea of a tree:
 - Removing the newly introduced `CarouselItem.DepthLayer` and/or computing it appropriately based on how deep an item is in the hierarchy.
 - Making the constant X offsets applied in [design PR](https://github.com/ppy/osu/pull/31774/files#diff-8f76f506418ce0d4adf726cfa9cac786cae298b17611913543e79bc2079d2433R30-R32) become based on the `CarouselItem`'s depth in the hierarchy instead.
 - Allowing `BeatmapCarousel` to easily determine whether to choose a small beatmap difficulty panel or a "standalone" beatmap single-diff panel, by simply checking whether the beatmap carousel item is a parent of a `BeatmapSetPanel` model or a `GroupDefinition`/no parent.

Of course all three points above can be solved by totally different means, but I found a basic tree model approach to be the natural solution, so I may as well raise it up for discussion.

peppy replied:
It's something i've considered and will try in the future. It has potential to optimise things if we hit a brick wall with the current structure.

So far the code complexity with flat list is very manageable and I'm not sure it can be maintained with tree.

bdach replied:
For reference I was party to an irl conversation on this exact question in December and my vague unsubstantiated feeling was also that a flat list was going to be more manageable than a hierarchical model - flat list does make some things more complicated, but also it will make some less so (e.g. I don't really want to think how the management of Y would look position if the backing model was no longer a flat list but hierarchical instead).

So far I see no strong indicators to change my opinion on this.

frenzibyte reviewed (commented):
Aside from a general thought I had on part of the current structure of carousel (which is, again, not a blocker), the changes in this PR look pretty sane to my eyes.

peppy commented:
Fixing the scrolling will be a separate endeavour.

Group traversal to me follows a simple rule: in the normal case you are using it to traverse sets. It should traverse sets disregarding other grouping.

When you interrupt it, in the case it's not at a difficulty, it will mostly behave like a "select" press until it reverts to the standard behaviour.

This is how stable works, and personally I think it feels good once you are used to it.

The only thing i'm neutral on is the changed behaviour when at a set header because this case doesn't exist in stable and is a bit undefined. The resultant behaviour from this PR makes it align with "just being a select press" if you can get behind that.

bdach reviewed (approved):
in that case there's only one thing left to do I guess
